### ISSUE TITLE
Increment attribute in `mergeEdge`

### ISSUE THREAD

**Issue by mef** [graphology/graphology#125]
I would like to know whether it is possible to increment attribute values when using `mergeEdge`.

### Use case

I am getting data from a realtime API, and need to update a graph on the fly:
* add edge with weight 1 if it does not exit already
* increment weight in case the edge exists.

I was thinking to use `mergeEdge`for this purpose. [This comment](https://github.com/graphology/graphology/issues/2#issuecomment-252275533) describes a way how it would be possible, with a callback function for `mergeEdge`. However, it does not seem to be in place in the reference implementation.

Do you guys have any recommendation about how I may proceed?

Thanks,

**Comment by Yomguithereal** [comment]
Hello @mef, we indeed have this use case in mind but did not yet implement something to make it quite easy.

The real issue here is that we cannot/don't want to enforce standard attribute naming on the user. Which means, like in `networkx`, for instance, that we cannot rely on the fact that the weight attribute would actually be called `weight`. What's more, we don't want to "sacralize" weight and make it something else than an attributes because graph are complex and every discipline has its own concepts such as labeling, weight etc. Attributes are a good & generic way to make everyone happy. So basically, what I would recommend right now is just to use `has` methods like this:

```js
if (graph.hasEdge(A, B))
  graph.updateEdgeAttribute(A, B, 'weight', x => x + 1);
else
  graph.addEdge(A, B, {weight: 1});
```

This said, we were speaking, with @boogheta, to add supplementary classes, somewhere outside of the reference implementation, like a library, to deal with this really common use case. Then you would have a tiny layer that would expose, e.g., a `WeightedGraph` class, that would overload the standard graph's method to modify the behavior of `addEdge` & `mergeEdge`. What is your opinion on this?

**Comment by Yomguithereal** [comment]
This said, maybe functions would be easier than full-fledged classes. Something like `addWeightedEdge`. But we would lose the opportunity to use those classes' instances in already existing scenarios.

**Comment by mef** [comment]
Thanks for the reply.

The proposed solution to use `has` works for me, no problem.

A solution I had in mind to support the use case without forcing users to name their weight `weight` would be to allow passing a function instead of attribute values inside `mergeEdge`:

````javascript
 graph.mergeEdge('John', 'Martha', (attributes) => {
    // attributes contains the values in the existing edge, or is `undefined`
    if (typeof attributes !== 'undefined')
        attributes.myWeight++
    else
        // attributes is undefined, this is a new edge
        attributes = {myWeight: 1}
    
    return attributes
})
````

From usage perspective, this is not much better than using `has` as suggested above.

From performance perspective, the gains of such a solution are not clear to me, you will know better.

Bottom line: I'm not sure about a suitable design at the moment, but doing an `if (graph.hasEdge(A,  B))`check is OK for me at the moment :)

**Comment by Yomguithereal** [comment]
This is an elegant solution but I feel it is not more concise nor more readable than using an `if has` condition. What's more, I am afraid it would entice people to mutate `attributes` object and to bypass setter functions, hence shunting potential events. What's more, this solution does not yield any advantage when using a `graph` with already existing code bases such as parsers.

E.g., if you have a specific JSON parser, using a special class that modifies the semantics of `addEdge/mergeEdge` to increment internal weights would enable someone with repeating edges in the JSON format to use already developed logic.

**Comment by Yomguithereal** [comment]
The idea would ressemble this basically:

```js
var graph = new WeightedGraph(); // or new WeightedGraph('customWeightAttrName');

// Then
graph.addEdge(0, 1);
graph.addEdge(0, 1); // Doesn't throw

graph.getEdgeAttribute(0, 1, 'weight');
>>> 2
```
